### PR TITLE
IO and redirection overhaul

### PR DESCRIPTION
**Pull request by IanManske**
# Description
The PR overhauls how IO redirection is handled, allowing more explicit and fine-grain control over `stdout` and `stderr` output as well as more efficient IO and piping.

To summarize the changes in this PR:
- Added a new `IoStream` type to indicate the intended destination for a pipeline element's `stdout` and `stderr`.
- The `stdout` and `stderr` `IoStream`s are stored in the `Stack` and to avoid adding 6 additional arguments to every eval function and `Command::run`. The `stdout` and `stderr` streams can be temporarily overwritten through functions on `Stack` and these functions will return a guard that restores the original `stdout` and `stderr` when dropped.
- In the AST, redirections are now directly part of a `PipelineElement` as a `Option<Redirection>` field instead of having multiple different `PipelineElement` enum variants for each kind of redirection. This required changes to the parser, mainly in `lite_parser.rs`.
- `Command`s can also set a `IoStream` override/redirection which will apply to the previous command in the pipeline. This is used, for example, in `ignore` to allow the previous external command to have its stdout redirected to `Stdio::null()` at spawn time. In contrast, the current implementation has to create an os pipe and manually consume the output on nushell's side. File and pipe redirections (`o>`, `e>`, `e>|`, etc.) have precedence over overrides from commands.

This PR improves piping and IO speed, partially addressing #10763. Using the `throughput` command from that issue, this PR gives the following speedup on my setup for the commands below:
| Command                     | Before (MB/s) | After (MB/s) | Bash (MB/s) |
| --------------------------- | -------------:| ------------:| -----------:|
| `throughput o> /dev/null`   | 1169          | 52938        | 54305       |
| `throughput \| ignore`      | 840           | 55438        | N/A         |
| `throughput \| null`        | Error         | 53617        | N/A         |
| `throughput \| rg 'x'`      | 1165          | 3049         | 3736        |
| `(throughput) \| rg 'x'`    | 810           | 3085         | 3815        |

(Numbers above are the median samples for throughput)

This PR also paves the way to refactor our `ExternalStream` handling in the various commands. For example, this PR already fixes the following code:
```nushell
^sh -c 'echo -n "hello "; sleep 0; echo "world"' | find "hello world"
```
This returns an empty list on 0.90.1 and returns a highlighted "hello world" on this PR.

Since the `stdout` and `stderr` `IoStream`s are available to commands when they are run, then this unlocks the potential for more convenient behavior. E.g., the `find` command can disable its ansi highlighting if it detects that the output `IoStream` is not the terminal. Knowing the output streams will also allow background job output to be redirected more easily and efficiently.

# User-Facing Changes
- External commands returned from closures will be collected (in most cases):
  ```nushell
  1..2 | each {|_| nu -c "print a" }
  ```
  This gives `["a", "a"]` on this PR, whereas this used to print "a\na\n" and then return an empty list.

  ```nushell
  1..2 | each {|_| nu -c "print -e a" }
  ```
  This gives `["", ""]` and prints "a\na\n" to stderr, whereas this used to return an empty list and print "a\na\n" to stderr.

- Trailing new lines are always trimmed for external commands when piping into internal commands or collecting it as a value. (Failure to decode the output as utf-8 will keep the trailing newline for the last binary value.) In the current nushell version, the following three code snippets differ only in parenthesis placement, but they all also have different outputs:

  1. `1..2 | each { ^echo a }`
     ```
     a
     a
     ╭────────────╮
     │ empty list │
     ╰────────────╯
     ```
  2. `1..2 | each { (^echo a) }`
     ```
     ╭───┬───╮
     │ 0 │ a │
     │ 1 │ a │
     ╰───┴───╯
     ```
  3. `1..2 | (each { ^echo a })`
     ```
     ╭───┬───╮
     │ 0 │ a │
     │   │   │
     │ 1 │ a │
     │   │   │
     ╰───┴───╯
     ```

  But in this PR, the above snippets will all have the same output:
  ```
  ╭───┬───╮
  │ 0 │ a │
  │ 1 │ a │
  ╰───┴───╯
  ```

- All existing flags on `run-external` are now deprecated.

- File redirections now apply to all commands inside a code block:
  ```nushell
  (nu -c "print -e a"; nu -c "print -e b") e> test.out
  ```
  This gives "a\nb\n" in `test.out` and prints nothing. The same result would happen when printing to stdout and using a `o>` file redirection.

- External command output will (almost) never be ignored, and ignoring output must be explicit now:
  ```nushell
  (^echo a; ^echo b)
  ```
  This prints "a\nb\n", whereas this used to print only "b\n". This only applies to external commands; values and internal commands not in return position will not print anything (e.g., `(echo a; echo b)` still only prints "b").

- `complete` now always captures stderr (`do` is not necessary).

# After Submitting
The language guide and other documentation will need to be updated.

### CODE REVIEW

**Comment by fdncred** [comment]
wow, that's some great performance gains! nice work!

**Comment by sholderbach** [comment]
How relevant is this part after https://github.com/nushell/nushell/pull/12120 or how much additional complexity do we need to handle?

> * `Command`s can also set a `IoStream` override/redirection which will apply to the previous command in the pipeline. This is used, for example, in `ignore` to allow the previous external command to have its stdout redirected to `Stdio::null()` at spawn time. In contrast, the current implementation has to create an os pipe and manually consume the output on nushell's side. File and pipe redirections (`o>`, `e>`, `e>|`, etc.) have precedence over overrides from commands.

**Comment by IanManske** [comment]
Still relevant, the overrides allow a more efficient implementation (this is part of how the speedup is obtained). The overrides are also opt-in, so most commands don't need to worry about it.

File: crates/nu-cmd-lang/src/core_commands/lazy_make.rs [hunk 1]
```diff
@@ -86,10 +86,13 @@ impl Command for LazyMake {
             }
         }
 
+        // TODO: do we reset stdio here?
```
Comment (IanManske) [inline crates/nu-cmd-lang/src/core_commands/lazy_make.rs:89]: Not sure where external command output from lazy record closures should go.

**Comment by fdncred** [comment]
FYI to people watching, we agreed yesterday to land this as soon as @IanManske cleans up a few minor things.

File: crates/nu-command/tests/commands/do_.rs [hunk 2]
```diff
@@ -80,7 +80,7 @@ fn capture_error_with_too_much_stderr_not_hang_nushell() {
         let actual = nu!(
             cwd: dirs.test(), pipeline(
             r#"
-            do -c {sh -c "cat a_large_file.txt 1>&2"} | complete | get stderr
+            sh -c "cat a_large_file.txt 1>&2" | complete | get stderr
```
Comment (sholderbach) [inline crates/nu-command/tests/commands/do_.rs:83]: Without `do` they should probably located in a different place.

Comment (IanManske) [reply crates/nu-command/tests/commands/do_.rs]: Good catch!

**Comment by devyn** [comment]
Congrats! I'm excited about this one

**Pull request merged**